### DISCUSSION TITLE
Disabling of tab in copilot.vim doesn't work

### DISCUSSION THREAD

**Discussion by SaskiaKeil** [community/community#8105] [Copilot Conversations] [answered]
According to the documentation you can set 
```
imap <silent><script><expr> <C-I> copilot#Accept("\<CR>")
let g:copilot_no_tab_map = v:true
```
to disable the completion by tab and map it to another key.

Once I set `let g:copilot_no_tab_map = v:true` though, I don't get any recommendations anymore. This is independent of the remapping to a different key.

Let me know if you need any other information on this issue :)

**Comment by cj** [comment]
@SaskiaKeil `let g:copilot_no_tab_map = v:true` doesn't seem to work, I bet if you run `Copilot status` it says `Copilot: <Tab> map has been disabled or is claimed by another plugin`

**Reply by SaskiaKeil** [reply]
You're right, it does. But isn't that expected? I disable the tab key of copilot so I can use it in another plugin. Instead I want to use Ctrl-I to use the suggested change.

**Reply by hakkm** [reply]
@SaskiaKeil after the status message of disabled tab, i runned `:verrbose map <Tab>` it showed me that the conflict happend in this file `~/.local/share/nvim/site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua` then in that file i changed ` ['<Tab>'] = { '\t' },` to   `  ['<C-J>'] = { '\t' },` and without setting `copilot_no_tab_map` to true, it worked fine.

**Comment by SaskiaKeil** [comment]
Well, a bit embarrassing but it turns out that I remapped tab to tab as C-I represents the tab key -.-
No problem with copilot here 😬

**Reply by shriharip** [reply]
Sorry, could not understand you @SaskiaKeil . I am having same issue. No autocompletions, how do I solve it. thanks

**Reply by SaskiaKeil** [reply]
I remapped the auto completion to `Control-i`, which in Vim is the same as the tab key though.
`imap <silent><script><expr> <C-I> copilot#Accept("\<CR>")`

This didn't work at all, but since I now mapped it to C-Space it works perfectly fine:
```
imap <silent><script><expr> <C-e> copilot#Accept('\<CR>')
let g:copilot_no_tab_map = v:true
```

**Comment by lukamanitta** [comment]
Seems like this is still an issue, I'm getting it with the following config:

```
vim.cmd('imap <silent><script><expr> <C-j> copilot#Accept("")')
vim.cmd("let g:copilot_no_tab_map = v:true")
```

with `Copilot: <Tab> map has been disabled or is claimed by another plugin` after running `:Copilot status`.
The recommendations stop showing up either when `g:copilot_no_tab_map` is set to true, or when another plugin (in my case `nvim-cmp`) claims the Tab key.

**Reply by lukamanitta** [reply]
Note: setting `g:copilot_assume_mapped = v:true` looks like it fixes it, for me at least.

**Comment by shaypal5** [comment]
Just passing through to say, wanting to work with `option+Tab` for my CoPilot completion (and `Tab` alone is used in a super-Tab like manner to choose between various nvim-cmp completions), `g:copilot_assume_mapped = v:true` solved it for me as well.

**Comment by bose-aritra2003** [comment]
Use this:
```
vim.g.copilot_no_tab_map = true
vim.api.nvim_set_keymap("i", "<C-J>", 'copilot#Accept("<CR>")', { silent = true, expr = true })
```
I also find it useful to disable copilot for certain filetypes:
  ```
vim.g.copilot_filetypes = {
    ["*"] = false,
    ["javascript"] = true,
    ["typescript"] = true,
    ["lua"] = false,
    ["rust"] = true,
    ["c"] = true,
    ["c#"] = true,
    ["c++"] = true,
    ["go"] = true,
    ["python"] = true,
  }
```
Hope that helps!

**Reply by shaypal5** [reply]
Hey< I'm very new to neovim. I've been using vim for around 5 years, but just moved to neovim, and I have no idea where to put the
```vim.api.nvim_set_keymap("i", "<C-J>", 'copilot#Accept("<CR>")', { silent = true, expr = true })```
line.

Can I get some context, please? :)

**Reply by shaypal5** [reply]
I've managed. Thank you. :)

**Reply by bose-aritra2003** [reply]
If this answer helped you, please do take a moment to upvote it.

**Reply by lminer** [reply]
@shaypal5 I'm having the same issue. Where do you put it?

**Reply by shaypal5** [reply]
I've added the following three lines:
```lua
vim.api.nvim_set_keymap("i", "<C-J>", 'copilot#Accept("<CR>")', { silent = true, expr = true })
vim.api.nvim_set_keymap("i", "<C-H>", 'copilot#Previous()', { silent = true, expr = true })
vim.api.nvim_set_keymap("i", "<C-K>", 'copilot#Next()', { silent = true, expr = true })
```

Almost at the top of the following file:
`~/.config/nvim/lua/config.lua`

**Comment by ChmielewskiKamil** [comment]
For anyone looking to set GitHub Copilot in AstroNvim, here is how you do it:
In your `user/init.lua`:
1. Add the following two lines to your vim options:
```lua
 options = {
    opt = {
      -- some existing code in here ...
    },
    g = {
      -- some existing code in here ...
      copilot_no_tab_map = true,
      copilot_assume_mapped = true,
    },
},
```
2. Under your key mappings add following lines (I wanted to use `F1`, `F2`, `F3`):
```lua
mappings = {
  -- first key is the mode
  n = {
          -- some existing code in here ...
  },
  t = {
          -- some existing code in here ...
  },
  i = {
      ["<F1>"] = { 'copilot#Accept("<CR>")', silent = true, expr = true },
      ["<F2>"] = { 'copilot#Next()', silent = true, expr = true },
      ["<F3>"] = { 'copilot#Previous()', silent = true, expr = true },
  }
},
```
3. Confirm that Copilot is working by running `:Copilot status`
4. *If you are on MacOS, you can disable the function key's default properties using KarabinerElements.

**Reply by Gouldilocks** [reply]
This worked perfectly for me, I'm using nVim with AstroVim, and I'm on an arm-based Mac

**Reply by nburnett** [reply]
THANK YOU! This fixed my main issue but when I do the F1 Accept I get these extra characters `<80>@7` at the end of the completion. Are these because of the osx f1 properties that were mentioned and that's what KarabinerElements is used for?

**Reply by Gouldilocks** [reply]
Yeah I am not sure what that is for, I get that as well. I use CTR-D to accept, and I get a couple funky characters like that sometimes.

**Reply by ChmielewskiKamil** [reply]
> THANK YOU! This fixed my main issue, but when I do the F1 Accept, I get these extra characters `<80>@7` at the end of the completion. Are these because of the osx f1 properties that were mentioned, and that's what KarabinerElements is used for?

I remember I also had a problem with weird symbols being inserted, like the one you mentioned, `<80>@7`. Since it was some time ago when I set this up, I am not 100% sure what the fix was, but KarabinerElements is your best bet. 

Another thing I would check is your nvim keymappings with `:map` or `:help index`. Make sure that `F1` is not bound to something weird.

**Reply by ChmielewskiKamil** [reply]
Once you have the KarabinerElements go to the `Function Keys` section and either set the toggle button at the bottom to "use all F1, F2, etc. keys as standard function keys" or set the ones you need from something like `F11 -> volume_decrement` to `F11 -> F11`.

**Reply by nburnett** [reply]
🤔 Tried setting up KarabinerElements and configuring things like you mentioned but still seem to be getting <80>kD and/or <80>@7

**Reply by nburnett** [reply]
ahhh, haha, figured it out! Thanks @ChmielewskiKamil for all the help, ended up being something with the key mapping like you mentioned. 

Found the fix [here](https://github.com/orgs/community/discussions/29817).  specifically the **noremap=true** & the **replace_keycodes=false**

@Gouldilocks maybe this will help with your issue as well.

**Reply by HarryKitchener** [reply]
Thanks @nburnett you saved me an hour of hell consulting the docs and chat gpt constantly failing me.

I was having the exact same issue and this the trailing characters.

**Reply by nburnett** [reply]
Glad it helped!

**Comment by jose19961975** [comment]
Okay I got it

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by wseabra** [comment]
None of the keymaps here worked for me for some reason. I had to use copilot with tab to write me a keymap and that worked fine:
`vim.keymap.set('i', '<C-S>', '<C-R>=copilot#Accept("")<CR>', { desc = "Use copilot", remap = false })`

**Comment by emkademy** [comment]
None of the suggestions have worked for me. Do you have any idea what else I could try?

**Comment by gitdemonic** [comment]
I set the "set paste" in vim, the tab can't accept the suggestion, 
When I remove "set paste", the tab immediate work fine. 
If you have same problem, can remove "set paste" first.

**Comment by parikshith078** [comment]
I tried this `vim.g.copilot_no_tab_map = true`
`vim.api.nvim_set_keymap("i", "<C-J>", 'copilot#Accept("<CR>")', { silent = true, expr = true })`  
in LunarVim but it says `unknown function`. Can anyone help please

**Comment by greglearns** [comment]
Looking at their code at (on my box): `dotfiles/vim/.vim/bundle/copilot.vim/plugin/copilot.vim`, or on github `https://github.com/github/copilot.vim/blob/release/plugin/copilot.vim`, you can see how they are mapping keys already.
Note: M means Meta which is the alt key.
```
    if empty(mapcheck('<M-]>', 'i'))
      imap <M-]> <Plug>(copilot-next)
    endif
    if empty(mapcheck('<M-[>', 'i'))
      imap <M-[> <Plug>(copilot-previous)
    endif
    if empty(mapcheck('<M-Bslash>', 'i'))
      imap <M-Bslash> <Plug>(copilot-suggest)
    endif
```

Given that they already are mapping 3 keys that are near each other, adding a 4th to accept suggestions and then disabling tab worked for me.

I did this:
```
let g:copilot_no_tab_map = 1
imap <script><silent><nowait><expr> <M-;> copilot#Accept()
```
With this, you can just type "Alt \" (without the space) whenever you want copilot to suggest a completion, then "Alt [" or "Alt ]" to scroll through options, then "Alt ;" to accept it.

**Comment by bpo217** [comment]
I'm gonna share the one and only thing that worked for me on my MacOS. `no_tab_map` by itself did not actually disable tab completion for copilot. I could set a different keymap and that would work fine but tab would always be enabled no matter what I did. Until I finally tried this:
```
vim.g.copilot_assume_mapped = true 
vim.keymap.del('i', '<Tab>')
vim.g.copilot_no_tab_map = true
vim.keymap.set('i', '<C-y>', 'copilot#Accept("\\<CR>")', {
  expr = true,
  replace_keycodes = false
})
```
Something I have for sure verified: if I remove `no_tab_map` line and leave the `keymap.del`, tab accept is STILL enabled. If I remove the `keymap.del` line and leave the `no_tab_map` line tab accept is STILL enabled. 

I, for whatever reason, have to have both:
```
vim.keymap.del('i', '<Tab>')
vim.g.copilot_no_tab_map = true
```
Again, this is the ONLY thing that has worked for me short of using a different plugin such as Tabnine, which I did for a while (Made me realize I really appreciate a neovim plugin that DOES NOT preset keymaps to use it). It wasn't the end of the world but I was sort of unlearning using tab the way I have been using it forever and working around it. 

I literally just had this moment where I was like, I wonder if I can unset a keybinding after it's been set, discovered del in help, and tried adding that in on top of the no_tab_completion

Tab still does what it's supposed to now too, which is insert n spaces in my case.

**Reply by erdajt** [reply]
Thanks man, this worked for me after 20 minutes of looking.

I thought this issue would have been fixed after **3 Years!** but no...

**Reply by tpope** [reply]
> I thought this issue would have been fixed after **3 Years!** but no...

This discussion post is approaching 3 years old, but the issue you're talking about is a load order issue and came onto my radar much more recently. See https://github.com/orgs/community/discussions/118483. The latest version of Copilot.vim contains a workaround.

**Comment by tpope** [comment, marked as answer]
The behavior described here, where suggestions are disabled if Copilot.vim doesn't think you have a map, was removed a few months ago.
